### PR TITLE
DEP: drop support for Python 3.10

### LINKED ISSUE #16900: Request: drop python 3.10 support now (since it will not be needed for 7.0)
### What is the problem this feature will solve?

In a few recent PRs, it would have been handy to just rely on python >=3.11 feature (#16886 -- `Exception.add_note`, #16892 -- `warnings.catch_warnings(action=..., category=...)`. Since by the time of 7.0, we should be dropping support for python 3.10 anyway, how about going ahead and do that in `.dev` now?

### Describe the desired outcome

Less work-arounds.

### Additional context

https://scientific-python.org/specs/spec-0000/#2024---quarter-4, drop support for
> python 	3.10 	released Oct 2021

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

* Fixes #16900.

I initially just wanted to measure the diff of doing this but ended up doing all (?) the work needed, so might as well open a ~draft~ PR.

This would enable:

* #16603
* #16879

### FAQ

Q: When is this going to be released?
A: End of 2024. Please see https://github.com/astropy/astropy/wiki/Release-Calendar

Q: Why so soon? Python 3.10 EOL is not till Oct 2026.
A: This is due to https://github.com/astropy/astropy-APEs/blob/main/APE18.rst that follows https://scientific-python.org/specs/spec-0000/ that recommends, "Support for Python versions be dropped **3 years** after their initial release."

### Subpackage approvals

- [x] cosmology (nstarman)
- [x] io.fits (pllim)
- [x] modeling (nden)
- [x] time (talcroft)
- [x] units (mhvk, nstarman)
- [x] utils.iers (mhvk)
- [x] utils (mhvk)

### After merge

Repo admin needs to update branch protection rules for `main`.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
I also revived changes from #16680 (re: https://github.com/astropy/astropy/pull/16680#issuecomment-2211058229)

**Comment by mhvk** [comment]
One more thing: I just checked and you ended up removing all actual usage of `typing_extensions` - maybe a reason to drop it as a dependency? Though since it is in the special `typing` category, no strong incentive to do so.

**Comment by neutrinoceros** [comment]
Good point, let's clean that up while we're at it. I'll keep the `typing` extra target though, we never know when we'll need it again and it doesn't hurt if it's empty for some time.

**Comment by neutrinoceros** [comment]
might as well undraft this now

**Comment by pllim** [comment]
Thanks! I posted about this on Slack and also on astropy-dev mailing list, just in case. So let's wait a bit before merge and preferably with multiple approvals.

**Comment by pllim** [comment]
FYI -- I added a bit more details to the original post above due to some questions I have been getting about this PR.

**Review by nstarman — APPROVED** [review]
LGTM! Thanks.
Though if we don't get rid of `typing_extensions` we can add `override` decorators.

**Comment by mhvk** [comment]
> Though if we don't get rid of `typing_extensions` we can add `override` decorators.

I am a bit torn by this, but think generally we should not include anything that we don't use. It just has to be clear that the moment we do need it again, it is totally OK to add back the dependency.

**Comment by nstarman** [comment]
It's been a typing-only dependency thus far. Supporting `override` and other py3.12+ features means a runtime-dependency.

File: pyproject.toml
Comment (pllim) [inline pyproject.toml:42]: Is the bugfix pin really necessary?

File: docs/changes/16903.other.rst
Comment (pllim) [inline docs/changes/16903.other.rst:1]: Should we also mention all the other version bumps in `pyproject.toml`?

Any more compatibility code we can remove from those other bumps as well?

Do these bumps mean those older version do not have support for Python 3.11?

File: tox.ini
Comment (pllim) [inline tox.ini:87]: `pyarrow>=7.0.0` in `pyproject.toml` needs to be updated to be consistent with this change here.

**Review by pllim — COMMENTED** [review]
Thanks! I like the simplifications but I have some questions, as follow.

File: pyproject.toml
Comment (neutrinoceros) [reply pyproject.toml:42]: 1.23.0 and 1.23.1 didn't support Python 3.11

File: docs/changes/16903.other.rst
Comment (neutrinoceros) [reply docs/changes/16903.other.rst:1]: > Do these bumps mean those older version do not have support for Python 3.11?

yes. 

> Should we also mention all the other version bumps in pyproject.toml?

I don't think it's needed, since these bumps are logical implications ?

> Any more compatibility code we can remove from those other bumps as well?

None that I could find.

File: tox.ini
Comment (neutrinoceros) [reply tox.ini:87]: done

**Review by taldcroft — APPROVED** [review]
Looks fine for `time` (and `table` and `io.ascii` since they are not impacted).

**Review by nden — APPROVED** [review]
modeling looks good to me

**Comment by pllim** [comment]
io.fits changes are trivial and I can approve for that if needed after another rebase to resolve conflict. I will have to do admin override anyway and change the branch protection.

I have no heard any objection to this across all the communication channels, so maybe we can get it merged as soon as tomorrow.

**Comment by neutrinoceros** [comment]
conflict resolved !

**Comment by pllim** [comment]
Some jobs failed with weird errors so I rebased (again) for you to see if failures were transient and so I can do a final review. Hope you don't mind.

**Comment by neutrinoceros** [comment]
No worries, I'm off for the day, so make yourself at home 😄

**Review by pllim — APPROVED** [review]
I added a bit more clean-ups and approved for FITS also. But maybe you want to have one last look before I merge, @neutrinoceros ? Thanks!

**Comment by pllim** [comment]
Ooops looks like I broke something. I will fix it later. My bad.

File: .pre-commit-config.yaml
Comment (pllim) [inline .pre-commit-config.yaml:1]: I removed tomli here to be consistent but I can revert if this is wrong. cc @neutrinoceros @nstarman @eerovaher

File: astropy/io/ascii/ipac.py
Comment (pllim) [inline astropy/io/ascii/ipac.py:557]: Not sure what this comment is about but since CI is green, obviously nothing needs fixing anymore, @taldcroft or @hamogu ?

**Comment by pllim** [comment]
Hmm oldest deps is misbehaving. It pulls numpy==1.25.2 but I am following the new pins. Will have to investigate more.

File: astropy/io/ascii/ipac.py
Comment (mhvk) [reply astropy/io/ascii/ipac.py:557]: I think it means one can replace the `if` statement and its clause with
```
col.width = max(len(vals[i]) for vals in data_str_vals, default=0)
```

File: .pre-commit-config.yaml
Comment (nstarman) [reply .pre-commit-config.yaml:1]: That should be good. Python 3.11 has the built-in toml parser ``tomllib``

File: astropy/io/ascii/ipac.py
Comment (pllim) [reply astropy/io/ascii/ipac.py:557]: Thanks, @mhvk ! I added a commit with your suggestion and credited you as co-author.

**Comment by pllim** [comment]
@neutrinoceros , the unsolicited numpy upgrade makes no sense to me and I ran out of time tonight. If you cannot figure it out while I am sleeping tomorrow, please let me know and I can dig some more. Thanks!

**Comment by neutrinoceros** [comment]
I don't know exactly where it's coming from but I could establish that we're getting numpy `1.25.2` (instead of `1.23.2`) only if we combine `oldestdeps` with `alldeps`. I think this means at least one of our optional dependencies isn't pinned and requires more recent versions of numpy, but it's still weird that the tox doesn't raise an error when resolving the environment (unless it does it *in increments* via pip ?).
I'm tempted to propose we keep oldestdeps as minimal as can be (drop the `alldeps` factor), but I can still spend a couple minutes trying to figure out which dependency from `alldeps` is the culprit so we can at least make an informed decision.

**Comment by neutrinoceros** [comment]
I think I found the culprit `dask[array]` was unpinned and brought `numpy>=1.24` with it.
It still seems a bit fragile to combine `oldestdeps` with `alldeps` but we can discuss that in a follow up issue. For now I think the immediate problem is resolved.

**Review by astrofrog — APPROVED** [review]
Looks good to me but needs a (simple) rebase

**Comment by neutrinoceros** [comment]
thanks ! rebased !

File: .pre-commit-config.yaml
Comment (neutrinoceros) [reply .pre-commit-config.yaml:1]: Actually I think this is orthogonal: pre-commit isn't necessarily installed within one's dev environment and can be installed system-wise against a different (possibly older) version of Python.
I *think* we need to also specify our minimal requirement for Python in `.pre-commit-config.yaml` as illustrated in https://pre-commit.com/#pre-commit-configyaml---top-level

I'll revert this for the present PR but I'll open a separate one with the complete change.

Comment (pllim) [reply .pre-commit-config.yaml:1]: Ok, sorry for the scope creep. I did extra grep and then could not help myself. 😅

**Comment by pllim** [comment]
Oh no, another conflict. Can you please rebase again? Thanks!

File: .pre-commit-config.yaml
Comment (neutrinoceros) [reply .pre-commit-config.yaml:1]: no worries 😸

**Comment by neutrinoceros** [comment]
rebased ! (I think it's time we merge this)

File: docs/changes/16903.other.rst
Comment (eerovaher) [inline docs/changes/16903.other.rst:1]: As a user I would like optional and mandatory dependencies to be distinguished (we don't need to go into any details beyond that).

**Comment by pllim** [comment]
I updated `main` branch protection rules. Please let me know if anything is amiss after a rebase of any open PR. Thanks, all!

**Comment by neutrinoceros** [comment]
Thanks @pllim for all your help !

File: docs/changes/16903.other.rst
Comment (eerovaher) [reply docs/changes/16903.other.rst]: `tomli` and `typing_extensions` were only needed for development, so they should not have been mentioned in the change log at all.

Comment (neutrinoceros) [reply docs/changes/16903.other.rst]: I plan to follow up on this. Thanks !

**Pull request merged**
